Add ViewNode to simplify render node management

IceSentry opened the pull request:
# Objective

- When writing render nodes that need a view, you always need to define a `Query` on the associated view and make sure to update it manually and query it manually. This is verbose and error prone.

## Solution

- Introduce a new `ViewNode` trait and `ViewNodeRunner` `Node` that will take care of managing the associated view query automatically.
- The trait is currently a passthrough of the `Node` trait. So it still has the update/run with all the same data passed in.
- The `ViewNodeRunner` is the actual node that is added to the render graph and it contains the custom node. This is necessary because it's the one that takes care of updating the node.

---

## Changelog

- Add `ViewNode`
- Add `ViewNodeRunner`

## Notes

Currently, this only handles the view query, but it could probably have a ReadOnlySystemState that would also simplify querying all the readonly resources that most render nodes currently query manually. The issue is that I don't know how to do that without a `&mut self`.

At first, I tried making this a default feature of all `Node`, but I kept hitting errors related to traits and generics and stuff I'm not super comfortable with. This implementations is much simpler and keeps the default Node behaviour so isn't a breaking change

## Reviewer Notes

The PR looks quite big, but the core of the PR is the changes in `render_graph/node.rs`. Every other change is simply updating existing nodes to use this new feature.

## Open questions

~~- Naming is not final, I'm opened to anything. I named it ViewQueryNode because it's a node with a managed Query on a View.~~
~~- What to do when the query fails? All nodes using this pattern currently just `return Ok(())` when it fails, so I chose that, but should it be more flexible?~~
~~- Is the ViewQueryFilter actually necessary? All view queries run on the entity that is already guaranteed to be a view. Filtering won't do much, but maybe someone wants to control an effect with the presence of a component instead of a flag.~~
~~- What to do with Nodes that are empty struct? Implementing `FromWorld` is pretty verbose but not implementing it means there's 2 ways to create a `ViewNodeRunner` which seems less ideal. This is an issue now because most node simply existed to hold the query, but now that they don't hold the query state we are left with a bunch of empty structs.~~
- Should we have a `RenderGraphApp::add_render_graph_view_node()`, this isn't necessary, but it could make the code a bit shorter.

JMS55 commented:
> Naming is not final, I'm opened to anything. I named it ViewQueryNode because it's a node with a managed Query on a View.

I'd go with just ViewNode, or ViewNodeWrapper or something

> What to do when the query fails? All nodes using this pattern currently just return Ok(()) when it fails, so I chose that, but should it be more flexible?

Ok(()) is fine. The query failing isn't actually an error. It just means no views are relevant to that node, e.g. if the FXAA node exists but no views are using FXAA, then the node just has nothing to do, but that's not an error.

> Is the ViewQueryFilter actually necessary? All view queries run on the entity that is already guaranteed to be a view. Filtering won't do much, but maybe someone wants to control an effect with the presence of a component instead of a flag.

I'd remove it. There's no need to filter by the time it gets to the node, usually you filter earlier on in extract or prepare and stuff. Adding this just adds boilerplate and complexity for no benefit.

JMS55 commented:
Any reason you're not using ViewQueryNode for the main pass nodes?

IceSentry commented:
> Any reason you're not using ViewQueryNode for the main pass nodes?

I should probably have made this a draft PR, but I wanted to play with the idea a bit first before updating everything. That's also the advantage of this approach, it doesn't force it on every node, but there's nothing blocking it for the main pass node.

kurtkuehnert commented on crates/bevy_core_pipeline/src/core_2d/mod.rs:
```diff
@@ -64,7 +64,7 @@ impl Plugin for Core2dPlugin {
 
         let pass_node_2d = MainPass2dNode::new(&mut render_app.world);
-        let tonemapping = TonemappingNode::new(&mut render_app.world);
+        let tonemapping = ViewQueryNodeRunner::new(TonemappingNode::new(), &mut render_app.world);
```
I am not a fan of how this reads.
We could just return a `ViewQuerryNodeRunner` from the `new` method.
So `TonemappingNode::new(&mut render_app.world)` would just return a `ViewQuerryNodeRunner<TonemappingNode>`.

kurtkuehnert replied:
Or something like `ViewQuerryNodeRunner<TonemappingNode>::new(&mut render_app.world` to make the wrapping more explicit. This could also be automatically implemented with a default or from world impl for the `ViewNode`.

IceSentry replied:
Yeah, that's the direction I want to go. Every node is already created with a new fn that is essentially a from_world so I'm heavily considering just enforcing it for the ViewNode

IceSentry replied:
So, after playing with this. It would require either a bunch of empty FromWorld impl on ViewNodes or use the Default trait instead and lose the ability to create a node with World data.

Both approaches feel pretty subpar honestly.

IceSentry replied:
I ended up using the Default impl for FromWorld and now use `ViewNodeRunner::<Node>::from_world(&mut world)` everywhere. It's still a bit verbose but it does the job.

kurtkuehnert reviewed (commented):
I really like the direction of this PR. 
Doing the same for the resources in a followup PR would be great as well.

kurtkuehnert commented:
I second @JMS55's suggestion on naming: `ViewNode` for `ViewQueryNode` and `ViewNodeWrapper` for the `ViewQueryNodeRunner`.

IceSentry commented:
Renamed to ViewNode and ViewNodeRunner. I think I prefer Runner instead of Wrapper since it's a bit more explicit in my head.

I also removed the filter since we agree that it seems unnecessary.

IceSentry commented:
I implemented a bunch more nodes. I added a new question to the description.

Essentially, after this PR, most Nodes are just empty structs because they only held the view query state and a couple of them create an empty mutex that is initialized on the first run. I'm trying to impl FromWorld on the `ViewNodeRunner`, but now most nodes don't need the World to be created, which means that if we want to use `ViewNodeRunner::<Node>::from_world(&mut world)` to create all view nodes we also need a bunch of empty impl FromWorld. I'm not sure how to tackle this issue.

JMS55 commented on examples/shader/post_process_pass.rs:
```diff
@@ -99,33 +105,20 @@ impl Plugin for PostProcessPlugin {
-    }
+impl ViewNode for PostProcessNode {
+    // This is the query that will be executed on the View.
+    // You can query for any component that exist on a view.
```
```suggestion
    // You can query for any component that exists on a view.
```

nicopap reviewed (approved):
Excellent job. Useful abstraction, also helps document the patterns used for render graph nodes in bevy.

IceSentry commented on crates/bevy_core_pipeline/src/tonemapping/node.rs:
```diff
@@ -2,72 +2,61 @@ use std::sync::Mutex;
 
 impl FromWorld for TonemappingNode {
-    fn from_world(world: &mut World) -> Self {
+    fn from_world(_world: &mut World) -> Self {
         Self {
-            query: QueryState::new(world),
             cached_bind_group: Mutex::new(None),
             last_tonemapping: Mutex::new(None),
         }
     }
 }
```
Not sure honestly, I think my reasoning at the time was that they wouldn't work with a derive Default so I didn't change them. I'll push an update with the changes.

nicopap replied:
It should work with `derive(Default)` since `Option<T>` implements `Default` regardless of T\[[1]\]. Also `Mutex<T>`  does for `T: Default`\[[2]\]. Also just tested it myself, it does compile.

[1]: https://doc.rust-lang.org/stable/std/option/enum.Option.html#impl-Default-for-Option%3CT%3E
[2]: https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#impl-Default-for-Mutex%3CT%3E

nicopap commented on crates/bevy_core_pipeline/src/tonemapping/node.rs:
```diff
@@ -2,72 +2,61 @@ use std::sync::Mutex;
 
 use super::{get_lut_bindings, Tonemapping};
 
 pub struct TonemappingNode {
```
It should work with `derive(Default)` since `Option<T>` implements `Default` regardless of T\[[1]\]. Also `Mutex<T>`  does for `T: Default`\[[2]\]. Also just tested it myself, it does compile.

[1]: https://doc.rust-lang.org/stable/std/option/enum.Option.html#impl-Default-for-Option%3CT%3E
[2]: https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#impl-Default-for-Mutex%3CT%3E

```suggestion
#[derive(Default)]
pub struct TonemappingNode {
```

IceSentry replied:
Oh yeah, it does, not sure what was wrong last time then

cart reviewed (approved):
This looks great! 
Just pushed some (very) minor tweaks:
* In the ViewNode interface, ROQueryItem can just be QueryItem for simplicity. The "read-onlyness" is implied by the ReadOnlyWorldQuery constraint
* I took this as a chance to use `&'static` everywhere (for consistency), rather than mixing and matching with `Read<>`
* Renamed ViewWorldQuery to ViewQuery to match the variable name and reduce "concept noise". I think "world-query-ness" can be implied here.
